microsoft/PowerToys
# Refactored Keyboard Manager UI code and added unit tests for Keyboard Manager UI

## Summary of the Pull Request

_What is this about?_

This PR refactors the methods used validation logic in the UI such that they are not linked to UI components, and adds 67 unit tests for those methods to cover scenarios arising in the UI by interacting with various controls.

## PR Checklist
* [X] Applies to #3114 

## Info on Pull Request

_What does this include?_

This PR moves the following methods to a single header file (`LoadingAndSavingRemappingHelper`) so that they can be unit tested:
- `CheckIfRemappingsAreValid` - used while clicking OK on the Remap key and shortcut UI windows to check if the remapping table is valid.
- `GetOrphanedKeys` - used to get the list of keys that have been orphaned with the current set of remappings while clicking OK on the Remap key window
- `PreProcessRemapTable` - used to combine Left and Right versions of all the modifiers if they are remapped to the same targets
- `ApplySingleKeyRemappings` - used to move the key remappings from the buffer to the keyboard manager state. This was earlier inside the `ApplyRemappings` lambda in the EditKeyboardWindow file
- `ApplyShortcutRemappings` - used to move the shortcut remappings from the buffer to the keyboard manager state. This was earlier inside the `ApplyRemappings` lambda in the EditShortcutWindow file
No logic changes have been made for the above methods, they have only been moved to another file. 23 tests were added for the above set of functions to take into account the different types of argument states.
![image](https://user-images.githubusercontent.com/32061677/89558918-c020e100-d7c9-11ea-8d9a-1be1bc984d9d.png)

This PR refactors the key table and shortcut table validation logic to methods in a separate header file (`BufferValidationHelpers`) such that both the methods no longer require any UI elements arguments to be passed. Instead the UI elements are used to compute vector representations for the current set of selections in the drop downs and this was passed as an argument to the validate method.
- `GetKeysFromStackPanel` was replaced with `GetSelectedIndicesFromStackPanel` which returns a vector of the indices selected in the set of drop downs. `GetKeyCodesFromSelectedIndices` was added to Helpers, such that `GetKeyCodesFromSelectedIndices` along with `GetSelectedIndicesFromStackPanel` can give the same result as `GetKeysFromStackPanel`.
- `CheckRepeatedModifier` which earlier used UI elements as argument was moved to Helpers and refactored so that it takes the list of keys directly rather than calling `GetKeysFromStackPanel`.
- The single key remap validation logic in the selection handler was moved to `ValidateAndUpdateKeyBufferElement`. This method validates the row as well as makes changes to the buffer based on the selection
- The basic shortcut remap validation logic in `ValidateShortcutSelection` was moved to `ValidateShortcutBufferElement`.  As part of the refactoring to make it independent of the UI elements, the Add drop down and step for clearing the unused drop downs was moved out of this function. Instead a variable is return which stores the error type and the action that needs to be done, and this action is process in `ValidateShortcutSelection` since these actions need to interact with the UI. This only validates the row, but does not update the buffer, since that is done in `SetSelectionHandler` and is more tied to the UI, so it cannot be unit tested along with the above. 
- Added 44 tests for `ValidateAndUpdateKeyBufferElement` and `ValidateShortcutBufferElement` to cover all the warning cases and normal entry cases. This includes setting a drop down to action key/modifier key/Null(blank)/None and checking if a warning needs to be shown and if an action needs to be done (where action could be deleting a drop down, adding a drop down or clearing blank drop downs). The tests cover most of the scenarios occurring when one drop down is changing. Some scenarios have not been included  (which are very similar to some of the other cases)  since it would bloat the test file size (since its already 3000 lines of code).
- To avoid bloating the test file size, some cases were combined into a single test, if they had the same drop down selection resulting in the same return values. For example, if selecting None on a dropdown resulted in the same return value irrespective of whether it was in the first/second column or if it was a hybrid column, these 3 scenarios were all added as separate Act/Assert lines in the same test.
![image](https://user-images.githubusercontent.com/32061677/89560032-4853b600-d7cb-11ea-8d81-c699c85a8e96.png)

## Validation Steps Performed

_How does someone test & validate?_

- Run the tests locally
- Run KBM and try adding and removing remappings with "Type" or drop downs and apply the remappings.
- After saving remappings, click the Remap button again to ensure the remappings load correctly.
